File: cpp/05_fichiers/lire.cpp
/*
 * MODULE 05 - Lire un fichier ligne par ligne avec std::ifstream + std::getline
 * =============================================================================
 * On ouvre un fichier en LECTURE, on vérifie qu'il existe bien (.is_open()),
 * puis on lit chaque ligne avec std::getline et on l'affiche. Le fichier se
 * referme tout seul à la fin (RAII).
 *
 * ⚠️ Lance d'abord ecrire.cpp pour créer exemples/notes.txt, et lance ce
 * programme DEPUIS LA RACINE du dépôt.
 *
 * Compiler puis lancer (depuis la racine) :
 *     g++ -Wall -std=c++17 cpp/05_fichiers/lire.cpp -o lire
 *     ./lire
 *
 * 🗺️ CHEMINEMENT DU PROGRAMME (les grandes étapes, dans l'ordre) :
 *    1. INCLUDE   : iostream (affichage), fstream (fichiers), string (les lignes).
 *    2. OUVRIR    : un std::ifstream sur "exemples/notes.txt".
 *    3. VÉRIFIER  : .is_open() — si le fichier n'existe pas, on prévient et on s'arrête.
 *    4. LIRE      : boucle while + std::getline = une ligne à la fois jusqu'à la fin.
 *    5. AFFICHER  : chaque ligne lue est envoyée à std::cout.
 *    6. FIN       : le fichier se referme automatiquement (RAII).
 */

#include <iostream>   // std::cout
#include <fstream>    // std::ifstream (flux pour lire un fichier)
#include <string>     // std::string (pour stocker chaque ligne)

int main() {
    // (2) On ouvre le fichier en LECTURE.
    std::ifstream fichier("exemples/notes.txt");

    // (3) Le fichier existe-t-il et a-t-il pu être ouvert ?
    //     Le « ! » se lit « PAS » : ici « si le fichier n'est PAS ouvert ».
    if (!fichier.is_open()) {
        std::cout << "Impossible d'ouvrir exemples/notes.txt." << std::endl;
        std::cout << "As-tu lance ./ecrire d'abord, depuis la racine du depot ?" << std::endl;
        return 1;   // on s'arrête proprement avec un code d'erreur
    }

    // (4 + 5) On lit le fichier LIGNE PAR LIGNE.
    //     std::getline met la prochaine ligne dans 'ligne' et renvoie le flux.
    //     Quand il n'y a plus rien à lire, la condition devient fausse : on sort.
    std::string ligne;
    while (std::getline(fichier, ligne)) {
        std::cout << ligne << std::endl;   // on affiche la ligne qu'on vient de lire
    }

    // (6) Le fichier se referme automatiquement quand 'fichier' disparaît (RAII).
    return 0;
}

File: go/05_fichiers/lire.go
/*
 * MODULE 05 - LIRE un fichier LIGNE PAR LIGNE en Go
 * =================================================
 * Montre la lecture morceau par morceau (ici, ligne par ligne) avec
 * os.Open + bufio.NewScanner. On ouvre le fichier, on le referme avec defer,
 * puis on parcourt les lignes une à une. Et si le fichier n'existe pas ?
 * os.Open renvoie une error : on la teste TOUT DE SUITE (réflexe du module 04).
 *
 * ⚠️ Ce fichier a SON PROPRE func main : on le lance SEUL.
 * Lancer DEPUIS LA RACINE du dépôt (après avoir lancé ecrire.go) :
 *     go run go/05_fichiers/lire.go
 *
 * 🗺️ CHEMINEMENT DU PROGRAMME (ce que main fait, dans l'ordre)
 *   1. Ouvrir "exemples/notes.txt" avec os.Open.
 *   2. Tester l'erreur : si le fichier n'existe pas, on prévient et on s'arrête.
 *   3. Prévoir de le refermer à la fin, quoi qu'il arrive (defer Close).
 *   4. Créer un "scanner" qui sait découper le fichier en lignes.
 *   5. Parcourir les lignes une par une et les afficher (numérotées).
 */

package main

import (
	"bufio" // boîte à outils pour lire efficacement (le Scanner de lignes)
	"fmt"   // boîte à outils d'affichage (Println, Printf)
	"os"    // boîte à outils "système" : ouvrir des fichiers...
)

func main() {

	// 1. OUVRIR le fichier en lecture (os.Open ne lit rien encore, il prépare).
	fichier, err := os.Open("exemples/notes.txt")

	// 2. TESTER l'erreur AVANT de s'en servir : si le fichier n'existe pas
	//    (ex. on n'a pas encore lancé ecrire.go), on le dit proprement et on sort.
	if err != nil {
		fmt.Println("Impossible d'ouvrir le fichier :", err)
		fmt.Println("Astuce : lance d'abord 'go run go/05_fichiers/ecrire.go'.")
		return
	}

	// 3. PRÉVOIR la fermeture : defer = "fais-le plus tard, juste avant de
	//    quitter main". Ainsi on n'oublie jamais de refermer le fichier.
	defer fichier.Close()

	// 4. CRÉER le scanner : par défaut, il découpe le fichier ligne par ligne.
	scanner := bufio.NewScanner(fichier)

	// 5. PARCOURIR : Scan() avance d'une ligne et renvoie true tant qu'il en reste.
	numero := 1
	for scanner.Scan() {
		ligne := scanner.Text() // le texte de la ligne courante (sans le \n)
		fmt.Printf("%d: %s\n", numero, ligne)
		numero++
	}
}
